### PR TITLE
Disable compatibility layer via hook

### LINKED ISSUE #8505: The `WooCommerce Product Grid Block` block is not visible on the front end when the `Products` block is on the same page.
## Describe the bug

The `WooCommerce Product Grid Block` block is not visible on the front end when the `Products` block is on the same page.

This issue is caused by the Compatibility Layer. Probably, the reason is it [removes default hooks](https://github.com/woocommerce/woocommerce-blocks/blob/f4ab91e00c307c798c05870ed427cc53b9e6818e/src/Templates/BlockTemplatesCompatibility.php/#L270). We should improve the logic and add the Compatibility Layer only when the `WooCommerce Product Grid Block` isn't on the page.

With #8507, we are disabling the Compatibility Layer because WC Blocks 9.6.0 will be included in WC Core, but we should on a "real" fix.

## To reproduce

1. With the [Twenty Twenty-Three](https://wordpress.org/themes/twentytwentythree/) theme installed, go to `Appearance » Editor`.
2. Edit the Product Catalog template, adding the Products block above the WooCommerce Product Grid Block.
3. Save.
4. Go to the front end and verify that the `WooCommerce Product Grid Block` isn't visible.

## Expected behavior

`WooCommerce Product Grid Block` should be visible. 

## Screenshots

https://user-images.githubusercontent.com/4463174/220608565-0912eae2-c337-40bf-b933-a37a20dd6669.mp4

### PR DESCRIPTION
**Pull request by gigitux**
This PR introduces a new filter called `woocommerce_disable_compatibility_layer` that allows disabling the compatibility layer.
At the same time, using the filter, the compatibility layer is disabled when Archive Product templates and Single Product template have the WooCommerce Product Grid Block.

Fixes #8505 

##### Archive Templates

1. With the [Twenty Twenty-Three](https://wordpress.org/themes/twentytwentythree/) theme installed, go to `Appearance » Editor`.
2. Edit the Product Catalog template, adding the Products block above the WooCommerce Product Grid Block.
3. Save.
4. Go to the front end and verify that the `WooCommerce Product Grid Block` is visible.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Add the `woocommerce_disable_compatibility_layer` filter to disable the compatibility layer. Disable the compatibility layer when Archive Product and Single Product templates contain the WooCommerce Product Grid Block.

### CODE REVIEW

File: src/Templates/BlockTemplatesCompatibility.php [hunk 1]
```diff
@@ -39,8 +39,52 @@ protected function init() {
...
 
-		add_filter( 'render_block_data', array( $this, 'update_render_block_data' ), 10, 3 );
-		add_filter( 'render_block', array( $this, 'inject_hooks' ), 10, 2 );
+		add_filter(
+			'render_block_data',
+			function( $parsed_block, $source_block, $parent_block ) {
+				/**
+				* Filters to disable the compatibility layer for the blockified templates.
+				*
+				* This hooks allows to disable the compatibility layer for the blockified.
```
Comment (Aljullu) [inline src/Templates/BlockTemplatesCompatibility.php]:
```suggestion
				* Filter to disable the compatibility layer for the blockified templates.
				*
				* This hook allows to disable the compatibility layer for the blockified templates.
```

File: src/BlockTemplatesController.php [hunk 2]
```diff
@@ -5,6 +5,7 @@
 use Automattic\WooCommerce\Blocks\Templates\BlockTemplatesCompatibility;
 use Automattic\WooCommerce\Blocks\Templates\ProductAttributeTemplate;
 use Automattic\WooCommerce\Blocks\Utils\BlockTemplateUtils;
+use Automattic\WooCommerce\Blocks\Templates\ProductSearchResultsTemplate;
```
Comment (Aljullu) [inline src/BlockTemplatesController.php]: Is this class used somewhere in this file? If not, do we need to `use` it?

**Review by Aljullu — APPROVED** [review]
One question and a suggestion to update some texts, but besides that, LGTM and is testing like a charm. :ok_hand:

File: src/BlockTemplatesController.php (hunk 2 above)
Comment (gigitux) [reply src/BlockTemplatesController.php]: Removed!

**Comment by gigitux** [comment]
> One question and a suggestion to update some texts, but besides that, LGTM and is testing like a charm. 👌

Thanks for the suggestions! I addressed your feedback! I will merge it 🚀

**Pull request merged**